tests/xtimer_usleep: improved test, added pin toggle

Josar opened the pull request:
Added `ERROR_US 500` to emphasize when errors occur.
Added `SLEEP_PIN ` and to probe sleep times at `SLEEP_GPIO_PIN`.

Atmega mcu will "hang" when the main ends, which will not matter here but should be addressed, imho.

kYc0o commented:
Can you explain more the purpose of this PR? I acknowledge for increasing the test cases, but I don't understand the need of the pin sleep.

Josar commented:
You can use an oscilloscope to see if the timer runbs as expected.
Or a seleae logic or other hardware, for hardware in the loop tests.

kYc0o commented:
Well the line you removed `// getchar();` is actually for that, it waits the test program to be launched to perform the checking through UART, so a pin is not needed and can be automatically tested. Did you checked the python script test under tests/?

Josar commented:
Ok, could you tell me how to start this test?

Still i think measuring with a scope could come handy when debuging.

kYc0o commented:
> Ok, so i run `make BOARD=jiminy-mega256rfr2 flash test`

Yes, that's enough to run the test.

> Still i think measuring with a scope could come handy when debuging.

Yes, however for the general use-case of the test it comes less handy, and even being optional, I think adding more code to the test may not help that much. However, if others have find it useful, I won't object.

@miri64 @kaspar030 do you have an opinion on this?

kYc0o commented on tests/xtimer_usleep/README.md:
```diff
@@ -1,8 +1,10 @@
+ and by providing capabilities to compare against an external timer.
+
+The sleep times can be probed at a pin if SLEEP_PIN` is set to 1 and the respective
+gpio pin is define as `SLEEP_GPIO_PIN`.
```
You missed a ` before SLEEP_PIN. You might give more hints on how to use it, e.g. add that it can be probed with an scope or something like that.

kYc0o commented on tests/xtimer_usleep/main.c:
```diff
@@ -24,35 +26,77 @@
-    puts("Please hit any key and then ENTER to continue");
-    getchar();
+    puts("Please hit any key and then ENTER to continue\n");
+    // getchar();
```
Please uncomment this line, it's needed for the automated test.

kYc0o commented on tests/xtimer_usleep/main.c:
```diff
@@ -24,35 +26,77 @@
            (unsigned)SLEEP_TIMES_NUMOF);
-    puts("Please hit any key and then ENTER to continue");
-    getchar();
+    puts("Please hit any key and then ENTER to continue\n");
```
`puts()` doesn't need any `\n`. Please remove.

kYc0o commented on tests/xtimer_usleep/main.c:
```diff
@@ -24,35 +26,77 @@
-            uint32_t diff, start;
-            start = xtimer_now_usec();
+
+            start_sleep = xtimer_now_usec();
```
Change the scope of your variables, they might be defined only in the context they're used.

kYc0o commented on tests/xtimer_usleep/main.c:
```diff
@@ -24,35 +26,77 @@
     }
     testtime = xtimer_now_usec() - start_test;
-    printf("Test ran for %" PRIu32 " us\n", testtime);
+    printf("\nTest ran for %" PRIu32 " us\n", testtime);
```
No need to add another `\n`.

kYc0o commented on tests/xtimer_usleep/main.c:
```diff
@@ -24,35 +26,77 @@
+    printf("\nTest ran for %" PRIu32 " us\n", testtime);
 
+    /* atmega mcu will hang if main ends */
+    while(1){}
```
This is another issue. This test is intended to run on any platform, and this is not needed.

kYc0o commented on tests/xtimer_usleep/main.c:
```diff
@@ -24,35 +26,77 @@
+            }else{
+                printf("Slept for %" PRIu32 " us (expected: %" PRIu32 " us)\n",
+                                   diff, sleep_times[n]);
+           }
```
Your `if - else` statements are very off on spacing, please stick to [coding conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions).

kYc0o commented on tests/xtimer_usleep/main.c:
```diff
@@ -24,35 +26,77 @@
-static const uint32_t sleep_times[] = { 10000, 50000, 100000 };
+static const uint32_t sleep_times[] = { 10000, 50000, 1234, 5678, 121212, 98765, 75000};
 
+#define ERROR_US 500
```
Errors of more than 500 microseconds are way too big IMHO. `xtimer_usleep()` is used for very short periods of time. If your hardware doesn't support it, it must be blacklisted for this test (e.g. running a very low speed timer).

kYc0o commented on tests/xtimer_usleep/main.c:
```diff
@@ -24,35 +26,77 @@
+    int32_t err;
+
+#if SLEEP_PIN
+    gpio_init( SLEEP_GPIO_PIN, GPIO_OUT );
```
Please remove the spaces after and before the parenthesis.

kYc0o commented on tests/xtimer_usleep/main.c:
```diff
@@ -24,35 +26,77 @@
+#if SLEEP_PIN
+#include "board.h"
+#include "periph/gpio.h"
+#define SLEEP_GPIO_PIN  GPIO_PIN(PORT_F, 7)
```
This should be surrounded by an `#ifndef` and the pin number should be passed from the outside (Makefile or command line) if you want to use it in this way. Hardcoded GPIO might lead to failures e.g. while flashing on certain devices. The default pin might be `GPIO_UNDEF`,  but still I'm not convinced this is a generic way to perform a test like this.

miri64 commented:
@Josar please provide fixup commits until the review says it's you can squash, so it is easier to track your changes.

miri64 commented:
Also, please specify in your commit message, what you did to the readme and how the tests were improved. A good style is to use an imperative style:

* `tests/xtimer_usleep: Update README to describe foo`
* `tests/xtimer_usleep: Add test cases to consider bar`

kaspar030 commented on tests/xtimer_usleep/main.c:
```diff
@@ -24,16 +26,33 @@
 
 #include "xtimer.h"
 #include "timex.h"
+#include "stdlib.h"
```
this is a standard header, so include ```<stdlib.h>``` and put above our headers.

kaspar030 commented on tests/xtimer_usleep/main.c:
```diff
@@ -43,12 +62,32 @@ int main(void)
+
+            diff = xtimer_now_usec() - start_sleep;
+
+            if(sleep_times[n] < diff - ERROR_US
```
this if needs some cleanup. parentheses, spaces, else on new line, ...

kYc0o commented on tests/xtimer_usleep/README.md:
```diff
@@ -1,10 +1,73 @@
+and the respective gpio `SLEEP_PORT` is defined in the makefile.
+
+CFLAGS += SLEEP_PIN=7
+CFLAGS += SLEEP_PORT=PORTF
```
The doc and the Makefile are not consistent on this.

kYc0o commented on tests/xtimer_usleep/README.md:
````diff
@@ -1,10 +1,73 @@
 ## Usage
+Executed from the project directory
+```
+make BOARD=<BoardName> flash test
````
Please avoid CamelCase.

kYc0o commented on tests/xtimer_usleep/README.md:
````diff
@@ -1,10 +1,73 @@
+```
+### On Error with terminal
+```
+
````
Why is this empty?

Josar commented:
@kYc0o  @kaspar030 good like this?

Josar commented:
@kaspar030 some further comments?

Josar commented:
@kYc0o squashed and rebased. Also changed the port initialization. Could you please test it again. 😅 
@kaspar030 could you have a second look and approve?

Josar commented:
@kaspar030 any further thoughts?

kYc0o commented on tests/xtimer_usleep/Makefile:
```diff
@@ -2,6 +2,12 @@ include ../Makefile.tests_common
+#FEATURES_REQUIRED += periph_gpio
+#CFLAGS += -DSLEEP_PIN=7
+#CFLAGS += -DSLEEP_PORT=PORT_F
+
```
Can you add just here `TEST_ON_CI_WHITELIST += all`

Josar commented:
@kYc0o @kaspar030 further requests?

kYc0o commented:
Everything goof for me. HIL tests passed.

Josar commented:
@kaspar030 any further requests?

kYc0o commented:
@Josar apparently you need to rebase this one to solve the conflicts.

kYc0o commented:
Great! It's green now. @kaspar030 can you confirm if your concerns were addressed to move on here?

kYc0o commented:
@cladmi if you want to test it would be awesome!

cladmi commented:
Test run correctly on IoT-LAB nodes:

* iotlab-m3: offset 14 us
* iotlab-a8-m3: offset 14 us
* wsn430-v1_3b: offset 135 us
* arduino-zero (samd21 cpu): offset 50us
* samr21-xpro: offset 44us

kYc0o commented:
@kaspar030 can you check if your concerns were addressed? This PR has 2 ACKs and I think we can merge it ASAP.
